src/mrgsolve.cpp

```diff
@@ -388,6 +388,19 @@ Rcpp::List EXPAND_OBSERVATIONS(
                             Rcpp::Named("index") = index);
 }
 
+Rcpp::List mat2df(Rcpp::NumericMatrix& x) {
```

Comment:
I'm not familiar with Rcpp df API. Could you walk me through this function? Thanks. @kylebaron

Reply:
The pattern basically follows `base::as.data.frame.matrix()`

1. Pass in a matrix
2. create a list with the same number of slots as the matrix has columns
3. Iterate through the matrix columns and copy into the list
4. `Rcpp::_` is just a way to get all the rows (or all columns)
5. The `rn(2)` bit replicates what happens in `base::.set_row_names`
6. Set the class to data frame
7. Add row names
8. Return

`as.data.frame.matrix()` does some other stuff but this does the essential steps for what I need.

Reply:
So `mat2df` doesn't take ownership of the matrix `x`'s data. Perhaps use `const` to ensure `x` is not modified? Like
```cpp
Rcpp::List mat2df(Rcpp::NumericMatrix const& x) {
```

Reply:
Sure.